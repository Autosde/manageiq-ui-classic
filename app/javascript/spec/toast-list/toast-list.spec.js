import React from 'react';
import { act } from 'react-dom/test-utils';
import fetchMock from 'fetch-mock';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import ToastList from '../../components/toast-list/toast-list';

describe('Toast list tests', () => {
  const initialState = {
    notificationReducer: {
      unreadCount: 1,
      isDrawerVisible: 'false',
      notifications: [
        {
          data: { link: 'http://localhost:3000/api/notifications/10000000003624' },
          href: 'http://localhost:3000/api/notifications/10000000003625',
          id: '10000000003625',
          message: 'Plan has completed with errors',
          notificationType: 'event',
          timeStamp: '2020-01-16T10:15:19Z',
          type: 'error',
          unread: true,
        },
      ],
      totalNotificationsCount: 0,
      toastNotifications: [
        {
          data: { link: 'http://localhost:3000/api/notifications/10000000003624' },
          href: 'http://localhost:3000/api/notifications/10000000003625',
          id: '10000000003625',
          message: 'Plan has completed with errors',
          notificationType: 'event',
          timeStamp: '2020-01-16T10:15:19Z',
          type: 'error',
          unread: true,
        },
      ],
      maxNotifications: 100,
    },
  };
  const mockStore = configureStore();

  beforeEach(() => {

  });

  afterEach(() => {
  });

  it('should render correctly with notifications', () => {
    const store = mockStore({ ...initialState });
    const wrapper = mount(
      <Provider store={store}>
        <ToastList />
      </Provider>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should not render without notifications', () => {
    const store = mockStore({
      ...initialState,
      notificationReducer: { ...initialState.notificationReducer, notifications: [], toastNotifications: [] },
    });
    const wrapper = mount(
      <Provider store={store}>
        <ToastList />
      </Provider>,
    );
    expect(wrapper.find('div.toast-notifications-list-pf')).toHaveLength(0);
  });

  it('should dispatch markNotificationRead after click on close button', async(done) => {
    fetchMock.postOnce('/api/notifications/10000000003625', {});
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <ToastList />
      </Provider>,
    );
    await act(async() => {
      wrapper.find('button.close.close-default').simulate('click');
    });
    const expectedPayload = {
      payload: '10000000003625',
      type: '@@notifications/markNotificationRead',
    };
    expect(store.getActions()).toEqual([expectedPayload]);
    done();
  });

  it('should dispatch removeToastNotification after delay', () => {
    jest.useFakeTimers();
    const store = mockStore({
      ...initialState,
      notificationReducer: {
        ...initialState.notificationReducer,
        toastNotifications: [
          {
            id: '10000000003625',
            type: 'success',
          },
        ],
      },
    });
    mount(
      <Provider store={store}>
        <ToastList />
      </Provider>,
    );
    const expectedPayload = {
      payload: '10000000003625',
      type: '@@notifications/removeToastNotification',
    };
    jest.runAllTimers();
    expect(store.getActions()).toEqual([expectedPayload]);
  });
});
