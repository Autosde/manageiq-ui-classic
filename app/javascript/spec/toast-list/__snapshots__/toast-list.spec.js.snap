// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast list tests should render correctly with notifications 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ToastList>
    <ToastNotificationList
      className=""
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="toast-notifications-list-pf"
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
      >
        <TimedToastNotification
          className=""
          key=".$10000000003625"
          onDismiss={[Function]}
          paused={false}
          persistent={true}
          timerdelay={8000}
          type="error"
        >
          <ToastNotification
            className=""
            onDismiss={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="error"
          >
            <Alert
              className="toast-pf"
              onDismiss={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="error"
            >
              <div
                className="alert toast-pf alert-danger alert-dismissable"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                <Button
                  active={false}
                  aria-hidden="true"
                  block={false}
                  bsClass="close"
                  bsStyle="default"
                  disabled={false}
                  onClick={[Function]}
                >
                  <button
                    aria-hidden="true"
                    className="close close-default"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <Icon
                      name="close"
                      type="pf"
                    >
                      <PatternflyIcon
                        className=""
                        name="close"
                      >
                        <span
                          aria-hidden="true"
                          className="pficon pficon-close"
                        />
                      </PatternflyIcon>
                    </Icon>
                  </button>
                </Button>
                <Icon
                  name="error-circle-o"
                  type="pf"
                >
                  <PatternflyIcon
                    className=""
                    name="error-circle-o"
                  >
                    <span
                      aria-hidden="true"
                      className="pficon pficon-error-circle-o"
                    />
                  </PatternflyIcon>
                </Icon>
                <span>
                  Plan has completed with errors
                </span>
              </div>
            </Alert>
          </ToastNotification>
        </TimedToastNotification>
        <TimedToastNotification
          className=""
          key=".$10000000003624"
          onDismiss={[Function]}
          paused={false}
          persistent={false}
          timerdelay={8000}
          type="success"
        >
          <ToastNotification
            className=""
            onDismiss={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="success"
          >
            <Alert
              className="toast-pf"
              onDismiss={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="success"
            >
              <div
                className="alert toast-pf alert-success alert-dismissable"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                <Button
                  active={false}
                  aria-hidden="true"
                  block={false}
                  bsClass="close"
                  bsStyle="default"
                  disabled={false}
                  onClick={[Function]}
                >
                  <button
                    aria-hidden="true"
                    className="close close-default"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <Icon
                      name="close"
                      type="pf"
                    >
                      <PatternflyIcon
                        className=""
                        name="close"
                      >
                        <span
                          aria-hidden="true"
                          className="pficon pficon-close"
                        />
                      </PatternflyIcon>
                    </Icon>
                  </button>
                </Button>
                <Icon
                  name="ok"
                  type="pf"
                >
                  <PatternflyIcon
                    className=""
                    name="ok"
                  >
                    <span
                      aria-hidden="true"
                      className="pficon pficon-ok"
                    />
                  </PatternflyIcon>
                </Icon>
                <span>
                  Plan has completed successfully
                </span>
              </div>
            </Alert>
          </ToastNotification>
        </TimedToastNotification>
      </div>
    </ToastNotificationList>
  </ToastList>
</Provider>
`;
